Add `split` Method to Subnet for Dividing Network Ranges into Smaller Subnets

alissn opened the pull request:
Hi,  

This pull request introduces a new `split` method to the `IPLib\Range\Subnet` class, enabling the division of a network range into smaller subnets. This feature supports both IPv4 and IPv6, making it versatile for various networking scenarios.  

### Usage  

```php
$subnet = \IPLib\Range\Subnet::parseString('192.168.112.203/24');
$smallerSubnets = $subnet->split(25);
```

### Result  

```php
[
    0 => IPLib\Range\Subnet {
        #fromAddress: IPLib\Address\IPv4 {
            #address: "192.168.112.0"
            #bytes: null
            #rangeType: null
            #comparableString: null
        }
        #toAddress: IPLib\Address\IPv4 {
            #address: "192.168.112.127"
            #bytes: null
            #rangeType: null
            #comparableString: null
        }
        #networkPrefix: 25
        #rangeType: null
    },
    1 => IPLib\Range\Subnet {
        #fromAddress: IPLib\Address\IPv4 {
            #address: "192.168.112.128"
            #bytes: null
            #rangeType: null
            #comparableString: null
        }
        #toAddress: IPLib\Address\IPv4 {
            #address: "192.168.112.255"
            #bytes: null
            #rangeType: null
            #comparableString: null
        }
        #networkPrefix: 25
        #rangeType: null
    }
]
```

### Key Benefits  
- **IPv4 and IPv6 Support**: Easily split both IPv4 and IPv6 ranges.  
- **Improved Efficiency**: Simplifies the process of working with smaller subnets from a larger range.  

Let me know if additional details or adjustments are needed!

mlocati reviewed (commented):
Why?

alissn commented:
Hi @mlocati,

Thank you for reviewing my pull request.

To handle large IP ranges efficiently, I used a generator class. This suggestion is from the IDE, recommending that the PHP version be updated to at least 5.5.

<img width="967" alt="image" src="https://github.com/user-attachments/assets/2a42f8e6-2825-416f-a4c7-56adfff9c787" />

mlocati commented:
Yel, generators require PHP 5.5. But I don't think I'd increase the minimum PHP version just for this reason

alissn commented:
@mlocati,

I have updated the code to return an array instead of a generator instance.

mlocati commented:
I'd
- remove the `RangeSplitInterface` interace and let any range class accept the split method (for "Range\Single" it will simply return an array with a single address)
- add tests

alissn commented:
Hi @mlocati,  

Do you have any feedback on this pull request?

mlocati commented:
> Do you have any feedback on this pull request?

@alissn The automated actions already gave you a feedback 😉 - see https://github.com/mlocati/ip-lib/pull/93/checks

alissn commented:
> @alissn The automated actions already gave you a feedback 😉 - see https://github.com/mlocati/ip-lib/pull/93/checks

@mlocati 
I have fixed the test.  

However, PHP **5.3, 5.4, and 5.5** are no longer relevant for me and many others.  
For more details, you can check the official PHP EOL page: [https://www.php.net/eol.php](https://www.php.net/eol.php).

mlocati commented:
@alissn I've updated the code of this pull request.
But your code only works if the range is a subnet, right? What about the other 2 types of ranges?

mlocati commented:
Also, it'd be nice to have a note about this new method in the README

alissn commented:
> @alissn I've updated the code of this pull request. 

Thanks 🙏  

> But your code only works if the range is a subnet, right? What about the other 2 types of ranges?

For pattern-ranges, this method works, but it requires making the `getNetworkPrefix` method public in the `Pattern` class.  

For single IP ranges, splitting doesn't have much meaning. However, I have updated the method to return a cloned instance of the original object.

alissn commented on src/Range/AbstractRange.php:
```diff
@@ -122,4 +124,49 @@ public function containsRange(RangeInterface $range)
+        $myNetworkPrefix = $this->getNetworkPrefix();
+        if ($networkPrefix === $myNetworkPrefix) {
+            return array(
+                new Subnet($this->getStartAddress(), $this->getEndAddress(), $networkPrefix),
```
@mlocati
I also suggest making this line `static` so the return type is based on the class.

Would you be okay with this change?

mlocati replied:
We have two options: thr split method always return Subnet instances, or it always returns instances of the range that's running the split method.
And I'd prefer this second option.
So: yes!

mlocati commented on src/Range/AbstractRange.php:
```diff
@@ -132,6 +132,10 @@ public function containsRange(RangeInterface $range)
      */
     public function split($networkPrefix)
     {
+        if ($this instanceof Single){
```
No need for clone (they are immutable).
Also, we should return a Subnet instance (or chance the phpdoc)

alissn replied:
i updated `phpdoc` on `RangeInterface`

mlocati commented on src/Range/Pattern.php:
```diff
@@ -310,7 +310,7 @@ public function getSize()
      * @return float|int
      */
-    private function getNetworkPrefix()
+    public function getNetworkPrefix()
```
Also "protected" should work

mlocati commented on src/Range/AbstractRange.php:
```diff
@@ -132,6 +132,10 @@ public function containsRange(RangeInterface $range)
      */
     public function split($networkPrefix)
     {
+        if ($this instanceof Single){
```
I'd instead override this method directly in the Single class

mlocati commented:
Also: please add some tests for the Single and Pattern classes (I don't think we need new test methods: adding cases to the two existing providers should be enough)

alissn commented:
@mlocati 
I have completed all the reviews.  

However, when passing a pattern-based range to the `split` method, it cannot return an array of `\IPLib\Range\Pattern` instances. Instead, I had to return `\IPLib\Range\Subnet` objects.

alissn commented:
```php
        $systemBitness = PHP_INT_SIZE * 8;
        $minPrefixByBitness = $maxPrefix - $systemBitness + 2;
        if ($minPrefixByBitness > $networkPrefix) {
            throw new OverflowException("The value of \$networkPrefix leads to too large ranges for the current machine bitness (you can use a value of at least {$minPrefixByBitness})");
        }
```

I removed this section because the `$networkPrefix` passed to the function cannot be smaller than the current range.  

Additionally, an error is already handled in the following section:  

```php
if ($networkPrefix < $myNetworkPrefix) {
    throw new OutOfBoundsException("The value of the \$networkPrefix parameter can't be smaller than the network prefix of the range ({$myNetworkPrefix})");
}
```

mlocati commented:
You shouldn't remove the chunk of code that throws an OverflowException: it handles the case when pow(2, ...) returns a non-integer value

alissn commented:
Reverted the change, but I couldn't add a test for it. If you can, please add a test case to ensure the exception is thrown.

mlocati commented:
> please add a test case to ensure the exception is thrown.

It's already there: https://github.com/mlocati/ip-lib/blob/410ce399f894b81bd839f1f59449800b58931236/test/tests/Ranges/RangesSplitTest.php#L588

alissn commented:
Based on the coverage report [here](https://coveralls.io/builds/72032538/source?filename=src%2FRange%2FAbstractRange.php#L154) this code path isn't covered by tests. 

![image](https://github.com/user-attachments/assets/40ef75e0-16e4-4b7e-88d1-e18dac9a56d4)

mlocati commented:
We send to Coveralls only one coverage, and it's for a case where we house a 64- bit PHP.
Maybe we can change that case to use a 32-bit PHP

mlocati commented:
Thanks!

mlocati commented:
See also #96 (which required #94 and #95)
